src/nvim/os/keyboard.c

```diff
@@ -0,0 +1,285 @@
+  }
+  return kCapsLock | kNumLock | kScrollLock;
+}
+
+#elif defined(__APPLE__)
+/// Get information about Num/Caps Lock state in MacOS
+///
+/// To be used in os_mods_status() function.
+///
+/// @param[out]  mods  Holds the information about the status of various locks
+///                    in form of a pointer to integer bitmask.
+/// @param[out]  err  Location where error message is to be saved
+///
+/// @return -1 in case of error and a mask specifying which values
+///         in mods are valid otherwise.
+static ModMask os_get_locks_status(ModMask *const mods, Error *const err)
+{
+  const CFMutableDictionaryRef mdict = IOServiceMatching(kIOHIDSystemClass);
+  const io_service_t ios = IOServiceGetMatchingService(kIOMasterPortDefault,
+                                                       (CFDictionaryRef)mdict);
+  if (!ios) {
+    if (mdict) {
+      CFRelease(mdict);
+    }
+    api_set_error(err, kErrorTypeException,
+                  "Unable to get service for default master port");
+    return -1;
+  }
+
+  io_connect_t ioc;
+  if (IOServiceOpen(ios, mach_task_self(), kIOHIDParamConnectType,
+                    &ioc) != kIOReturnSuccess) {
+    IOObjectRelease(ios);
+    api_set_error(err, kErrorTypeException,
+                  "Unable to get service for HID system class");
+    return -1;
+  }
+  IOObjectRelease(ios);
+
+  bool stateCaps;
+  if (IOHIDGetModifierLockState(ioc, kIOHIDCapsLockState,
+                                &stateCaps) != kIOReturnSuccess) {
+    IOServiceClose(ioc);
+    api_set_error(err, kErrorTypeException,
+                  "Unable to query CapsLock state");
+    return -1;
+  }
+
+  bool stateNums;
+  if (IOHIDGetModifierLockState(ioc, kIOHIDNumLockState,
+                                &stateNums) != kIOReturnSuccess) {
+    IOServiceClose(ioc);
+    api_set_error(err, kErrorTypeException,
+                  "Unable to query NumLock state");
+    return -1;
+  }
+  if (stateCaps) {
+    *mods |= kCapsLock;
+  }
+  if (stateNums) {
+    *mods |= kNumLock;
+  }
+  IOServiceClose(ioc);
+  return kCapsLock | kNumLock;
+}
+#else
+/// Dummy function used in case of no known platform detected.
+///
+/// @param[out]  mods  Holds the information about the status of various locks
+///                    in form of a pointer to integer bitmask.
+/// @param[out]  err  Location where error message is to be saved
+///
+/// @return Always return 0.
+static ModMask os_get_locks_status(ModMask *const mods, Error *const err)
+{
+  *mods = 0;
+  return 0;
+}
+#endif
+
+/// Get information about Num/Caps/Scroll Lock state.
+///
+/// To be used in nvim_get_keyboard_mods_state() function.
+///
+/// @param[out]  dict  Pointer to dictionary where information about modifiers
+///                    is to be dumped.
+/// @param[out]  err  Location where error message is to be saved
+///
+/// @return true in case of no error, false otherwise.
+bool os_mods_status(Dictionary *const dict, Error *const err)
```

Comment:
And `FUNC_ATTR_NONNULL_ALL` here (warning for unused result is not needed as result is actually redundant).